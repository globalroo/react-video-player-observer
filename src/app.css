.App {
	text-align: center;
}

.App-logo {
	animation: App-logo-spin infinite 20s linear;
	height: 60px;
}

.App-header {
	background-color: #222;
	padding: 10px;
	color: white;
}

.App-intro {
	font-size: medium;
}

@keyframes App-logo-spin {
	from {
		transform: rotate(0deg);
	}
	to {
		transform: rotate(360deg);
	}
}

.btn-group-max-width {
	width: 100%;
}

.now-playing {
	color: #00bc8c;
	text-transform: uppercase;
	font-weight: bold;
}

.player-observer-container {
	background-color: #333333;
	float: left;
	padding: 10px;
	padding-top: 20px;
	border: 1.5px solid #00bc8c;
	border-radius: 10px;
	min-width: 100%;
	box-shadow: 0px 0px 50px rgba(0,0,0,.6);
}

.panel {
	box-shadow: 0px 0px 50px rgba(0,0,0,.6);
}
