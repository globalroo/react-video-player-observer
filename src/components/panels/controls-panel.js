import React from "react";
import PropTypes from "prop-types";

export const createClickHandler = ({ loadAsset, asset }) => () => loadAsset(asset);

export const MovieButton = ({ asset, loadAsset }) => {
	const clickHandler = createClickHandler({ loadAsset, asset });
	return (
		<div className="btn-group" role="group">
			<button type="button" onClick={clickHandler} className="btn btn-info">{asset.name}</button>
		</div>
	);
};

MovieButton.propTypes = {
	loadAsset: PropTypes.func,
	asset: PropTypes.object
};

export const ControlsPanel = ({ availableMedia, loadAsset, asset }) => (
	<div className="panel panel-default">
		<div className="panel-heading">
			<h4>Controls</h4>
		</div>
		<div className="panel-body">
			<div className="btn-group btn-group-vertical btn-group-max-width" role="group" aria-label="Switch movies">
				{
					availableMedia.map((asset, ix) =>
						<MovieButton key={`asset_${ix}`} asset={asset} loadAsset={loadAsset} />)
				}
			</div>
		</div>
		<div className="panel-footer">
			{(asset && asset.name) && <small>Now playing: <span className="now-playing">{asset.name}</span></small>}
		</div>
	</div>
);

ControlsPanel.propTypes = {
	availableMedia: PropTypes.array,
	loadAsset: PropTypes.func,
	asset: PropTypes.object
};
