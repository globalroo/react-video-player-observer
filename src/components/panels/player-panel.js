import { Player } from "src/components/player/player";
import React from "react";
import PropTypes from "prop-types";

export const PlayerPanel = ({ onPlayerLoaded }) => (
	<div className="panel panel-default">
		<div className="panel-heading">
			<h4>Video Player</h4>
		</div>
		<div className="panel-body">
			<Player onPlayerLoaded={onPlayerLoaded} />
		</div>
	</div>
);

PlayerPanel.propTypes = {
	onPlayerLoaded: PropTypes.func,
};
