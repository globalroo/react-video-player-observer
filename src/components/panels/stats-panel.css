.latestEvent {
	font-size:130%;
}
.halfColumn {
	width: 50%;
}
