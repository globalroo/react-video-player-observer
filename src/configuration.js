export const HISTORICAL_EVENTS = 12;

export const getAvailableMedia = () => ([
	{
		name: "Sintel",
		poster: "http://media.w3.org/2010/05/sintel/poster.png",
		src: "https://media.w3.org/2010/05/sintel/trailer.mp4",
	},
	{
		name: "Big Buck Bunny - Trailer",
		poster: "http://media.w3.org/2010/05/bunny/poster.png",
		src: "http://media.w3.org/2010/05/bunny/trailer.mp4",
	},
	{
		name: "Big Buck Bunny - Movie",
		poster: "http://media.w3.org/2010/05/bunny/poster.png",
		src: "http://media.w3.org/2010/05/bunny/movie.mp4",
	},
]);

export const PLAYER_EVENTS_TO_OBSERVE = {
	ABORT: "abort",
	CANPLAY: "canplay",
	CANPLAYTHROUGH: "canplaythrough",
	DURATIONCHANGE: "durationchange",
	EMPTIED: "emptied",
	ENDED: "ended",
	ERROR: "error",
	LOADEDDATA: "loadeddata",
	LOADEDMETADATA: "loadedmetadata",
	LOADSTART: "loadstart",
	PAUSE: "pause",
	PLAY: "play",
	PLAYING: "playing",
	PROGRESS: "progress",
	RATECHANGE: "ratechange",
	RESIZE: "resize",
	SEEKED: "seeked",
	SEEKING: "seeking",
	STALLED: "stalled",
	SUSPEND: "suspend",
	TIMEUPDATE: "timeupdate",
	VOLUMECHANGE: "volumechange",
	WAITING: "waiting",
};
