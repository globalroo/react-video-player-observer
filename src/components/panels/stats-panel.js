import "./stats-panel.css";
import React from "react";
import PropTypes from "prop-types";

export const StatsPanel = ({ playerState, eventHistory }) => {
	const currentTime = (playerState && playerState.currentTime) ? playerState.currentTime.toFixed(3) : 0;
	const remainingTime = (playerState && playerState.remainingTime) ? playerState.remainingTime.toFixed(3) : 0;
	const stateClass = ( playerState.error.code !== 0) ? "danger" : "info";
	return (
		<div className="panel panel-default">
			<div className="panel-heading">
				<h4>Observer</h4>
			</div>
			<div className="panel-body">
				<table className="table table-striped table-responsive">
					<tbody>
						<tr>
							<td className="halfColumn">currentTime</td>
							<td className="halfColumn">{currentTime}</td>
						</tr>
						<tr>
							<td>remainingTime</td>
							<td>{remainingTime}</td>
						</tr>
						<tr>
							<td>error?</td>
							<td className={stateClass}>{playerState.error.description} [{playerState.error.code}]</td>
						</tr>
						{
							eventHistory.map(({ event, timeStamp }, ix) => {
								const className = (ix === 0) ? "success latestEvent" : "";
								return (
									<tr className={className} key={ix} >
										<td>{event}</td>
										<td>{timeStamp}</td>
									</tr>
								);
							})
						}
					</tbody>
				</table>
			</div>
		</div>
	);
};

StatsPanel.propTypes = {
	playerState: PropTypes.object,
	eventHistory: PropTypes.array
};
