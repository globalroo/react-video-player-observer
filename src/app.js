import "./app.css";

import React, { Component } from "react";
import PropTypes from "prop-types";

import { ControlsPanel } from "src/components/panels/controls-panel";
import { PLAYER_EVENTS_TO_OBSERVE, HISTORICAL_EVENTS } from "./configuration";
import { PlayerPanel } from "src/components/panels/player-panel";
import { StatsPanel } from "src/components/panels/stats-panel";
import { PlayerObserver } from "src/player-observer/player-observer";
import { getFormattedTime } from "src/utils/date-helper";

import logo from "./logo.svg";

const playerState = {
	currentTime: 0,
	remainingTime: 0,
	error: {
		code: 0,
		description: ""
	}
};

const selectedAsset = {
	name: "",
	poster: "",
	src: "",
};

export const initialState = {
	playerState,
	selectedAsset,
	eventHistory: []
};

export class PlayerApp extends Component {
	constructor(props) {
		super(props);
		this.state = initialState;
		this.eventListeners = [];
		this.observer = PlayerObserver.getInstance();
	}

	primePlayer(payload) {
		this._player.src = this.state.selectedAsset.src;
		this._player.poster = this.state.selectedAsset.poster;
	}

	onPlayerLoaded = (player) => {
		this._player = player;
		this.observer.initialisePlayer(player);
	}

	loadAsset = (selectedAsset) => {
		this.setState(
			() => ({ ...this.state, selectedAsset }),
			this.primePlayer
		);
	}

	addEventToHistory(event, oldEventHistory) {
		const eventHistory = [...oldEventHistory];
		const latestEvent = {
			timeStamp: getFormattedTime(new Date()),
			event,
		};
		if (eventHistory &&
			eventHistory.length === HISTORICAL_EVENTS) {
			eventHistory.pop();
		}
		eventHistory.unshift(latestEvent);
		return eventHistory;
	}

	handleState = (event, playerState) => {
		const eventHistory = this.addEventToHistory(event, this.state.eventHistory);
		const newState = { playerState, eventHistory };
		this.setState (newState);
	}

	subscribeEvents(trackEvents) {
		Object.keys(trackEvents).forEach((event) => {
			this.eventListeners.push(this.observer.subscribe(trackEvents[event], this.handleState));
		});
	}

	unsubscribeEvents() {
		this.eventListeners.forEach((eventUnsubscribeFn) => {
			eventUnsubscribeFn();
		});
	}

	componentDidMount() {
		this.subscribeEvents(PLAYER_EVENTS_TO_OBSERVE);
		this.loadAsset(this.props.availableMedia[0]);
	}

	componentWillUnmount() {
		this.unsubscribeEvents();
	}

	render() {
		return (
			<div className="App container">
				<div className="App-header col-xs-12">
					<img src={logo} className="App-logo" alt="logo" />
					<h4> React Player Observer Test Harness </h4>
				</div>
				<div className="player-observer-container">
					<div className="col-xs-12 col-sm-6">
						<PlayerPanel onPlayerLoaded={this.onPlayerLoaded} />
						<ControlsPanel asset={this.state.selectedAsset} loadAsset={this.loadAsset} availableMedia={this.props.availableMedia}/>
					</div>
					<div className="col-xs-12 col-sm-6">
						<StatsPanel playerState={this.state.playerState} eventHistory={this.state.eventHistory}/>
					</div>
				</div>
			</div>
		);
	}
}

PlayerApp.propTypes = {
	availableMedia: PropTypes.array,
};
