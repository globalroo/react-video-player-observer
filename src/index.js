import React from "react";
import ReactDOM from "react-dom";
import "./index.css";
import { PlayerApp } from "./app";
import registerServiceWorker from "./registerServiceWorker";

import { getAvailableMedia } from "./configuration";

const availableMedia = getAvailableMedia();

ReactDOM.render(
	<PlayerApp
		availableMedia={availableMedia}
	/>,
	document.getElementById("root")
);
registerServiceWorker();
