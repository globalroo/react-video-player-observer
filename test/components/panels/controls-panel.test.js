import React from "react";
import * as renderer from "react-test-renderer";
import { getAvailableMedia } from "src/configuration";
import { ControlsPanel, MovieButton, createClickHandler } from "src/components/panels/controls-panel";

describe("Controls Panel tests", () => {

	const availableMedia = getAvailableMedia();

	test("Controls panel renders as expected", () => {
		const asset = availableMedia[0];
		const component = renderer.create(
			<ControlsPanel
				asset={asset}
				loadAsset={() => {}}
				availableMedia={availableMedia}
			/>
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	test("MovieButton renders as expected", () => {
		const asset = availableMedia[0];
		const component = renderer.create(
			<MovieButton
				asset={asset}
				loadAsset={() => {}}
			/>
		);
		expect(component.toJSON()).toMatchSnapshot();
	});

	test("Click handler creates a HoC that calls the callback triggernig asset to change in UI", () => {
		const asset = availableMedia[0];
		const spyLoadAsset = jest.fn();
		const hocClickHandler = createClickHandler({asset, loadAsset: spyLoadAsset});
		hocClickHandler();
		expect(spyLoadAsset).toHaveBeenCalledWith(asset);
	});
});
