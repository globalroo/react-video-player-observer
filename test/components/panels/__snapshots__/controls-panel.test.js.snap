// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls Panel tests Controls panel renders as expected 1`] = `
<div
  className="panel panel-default"
>
  <div
    className="panel-heading"
  >
    <h4>
      Controls
    </h4>
  </div>
  <div
    className="panel-body"
  >
    <div
      aria-label="Switch movies"
      className="btn-group btn-group-vertical btn-group-max-width"
      role="group"
    >
      <div
        className="btn-group"
        role="group"
      >
        <button
          className="btn btn-info"
          onClick={[Function]}
          type="button"
        >
          Sintel
        </button>
      </div>
      <div
        className="btn-group"
        role="group"
      >
        <button
          className="btn btn-info"
          onClick={[Function]}
          type="button"
        >
          Big Buck Bunny - Trailer
        </button>
      </div>
      <div
        className="btn-group"
        role="group"
      >
        <button
          className="btn btn-info"
          onClick={[Function]}
          type="button"
        >
          Big Buck Bunny - Movie
        </button>
      </div>
    </div>
  </div>
  <div
    className="panel-footer"
  >
    <small>
      Now playing: 
      <span
        className="now-playing"
      >
        Sintel
      </span>
    </small>
  </div>
</div>
`;

exports[`Controls Panel tests MovieButton renders as expected 1`] = `
<div
  className="btn-group"
  role="group"
>
  <button
    className="btn btn-info"
    onClick={[Function]}
    type="button"
  >
    Sintel
  </button>
</div>
`;
